# **Task 4.6**

## 01. Importing Libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Data

In [2]:
# Creating a path to main project folder
path = r'C:\Users\jboer\OneDrive\Documents\Career Foundry\Instacart Basket Analysis'

In [3]:
# Importing products_checked.csv file
df_prods = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'), index_col = False)

In [4]:
# Importing orders_products_combined.pkl file
df_ords_prods = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_combined.pkl'))

## 03. Checking Data

In [5]:
# Checking shape of df_ords_prods to make sure it's the same as the one exported
df_ords_prods.shape

(32434489, 11)

##### Confirmed that the imported "orders_products_combined.pkl" file contains the same number of rows of the one I exported: 32,434,489 rows and 11 columns.

In [6]:
# Checking shape of df_prods
df_prods.shape

(49672, 6)

In [7]:
# Checking for common key to merge in df_ords_prods
df_ords_prods.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     0.0         196                  1   
1                  8                     0.0       14084                  2   
2                  8                     0.0       12427                  3   
3                  8                     0.0       26088                  4   
4                  8                     0.0       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [8]:
# Checking for common key to merge in df_prods
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

### 03.1. Dropping Unnecessary Columns

In [18]:
# Dropping "Unnamed: 0" and "_merge" columns from df_ords_prods
df_ords_prods = df_ords_prods.drop(columns = ['Unnamed: 0', '_merge'])

In [19]:
# Checking success of dropping columns in df_ords_prods
df_ords_prods.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     0.0         196                  1          0  
1                     0.0       14084                  2          0  
2                     0.0       12427                  3          0  
3                     0.0       26088                  4          0  
4                     0.0       26405                  5          0

In [20]:
# Dropping "Unnamed: 0" column from df_prods
df_prods = df_prods.drop(columns = ['Unnamed: 0'])

In [21]:
# Checking success of dropping column in df_prods
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

##### The common column to be used as a key to merge the two dataframes is the "product_id" column.

## 04. Merging Data

In [23]:
# Testing inner join merge with key "product_id"
pd.merge(df_ords_prods, df_prods, on = 'product_id', indicator = True)

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32404854   2977660   206209            13                   1   
32404855   2977660   206209            13                   1   
32404856   2977660   206209            13                   1   
32404857   2977660   206209            13                   1   
32404858   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     0.0         196   
1                         8                     0.0       14084   
2                         8                     0.0       12427   
3                         8                     0.0       26088   
4                         8                     0.0       26405   
...                     ...                     ...         ...   
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered  \
0                         1          0   
1                         2          0   
2                         3          0   
3                         4          0   
4                         5          0   
...                     ...        ...   
32404854                  5          1   
32404855                  6          0   
32404856                  7          0   
32404857                  8          0   
32404858                  9          0   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32404854                             Tomato Paste         9              9   
32404855          Brownie Crunch High Protein Bar         3             19   
32404856    High Protein Bar Chunky Peanut Butter         3             19   
32404857      Chocolate Peanut Butter Protein Bar         3             19   
32404858      Roasted & Salted Shelled Pistachios       117             19   

          prices _merge  
0            9.0   both  
1           12.5   both  
2            4.4   both  
3            4.7   both  
4            1.0   both  
...          ...    ...  
32404854     5.6   both  
32404855     5.9   both  
32404856     4.2   both  
32404857     4.9   both  
32404858     1.7   both  

[32404859 rows x 14 columns]

##### Test merge was successful. Now will merge both datatframes into a new dataframe.

In [24]:
# Merging df_prods with df_ords_prods using key "product_id"
ords_prods_merge = df_ords_prods.merge(df_prods, on = 'product_id', indicator = True)

In [25]:
# Checking new merged dataframe ords_prods_merge
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [26]:
# Checking shape of ords_prods_merge
ords_prods_merge.shape

(32404859, 14)

In [27]:
# Confirming results of merge by checking "_merge" frequency
ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [28]:
# Creating outer join merge with key "product_id" to test "_merge" frequency
ords_prods_merge_outer = df_ords_prods.merge(df_prods, on = 'product_id', how = 'outer', indicator = True)

In [29]:
# Comparing results of inner and outer merge by checking outer "_merge" frequency
ords_prods_merge_outer['_merge'].value_counts()

_merge
both          32404859
left_only        30200
right_only          11
Name: count, dtype: int64

##### As we can see from comparing the inner and outer merges frequencies, there is not a full match in the inner merge because we can see there are 30,200 counts of left-only and 11 counts of right-only in the outer merge. So, those were left out of the inner merge.

In [30]:
# Checking for missing values in ords_prods_merge_outer
ords_prods_merge_outer.isnull().sum()

order_id                     11
user_id                      11
order_number                 11
orders_day_of_week           11
order_hour_of_day            11
days_since_prior_order       11
product_id                    0
add_to_cart_order            11
reordered                    11
product_name              30200
aisle_id                  30200
department_id             30200
prices                    30200
_merge                        0
dtype: int64

##### After checking the missing values in the outer merge, I will use the inner merge dataframe moving forward. Most of the columns that have missing values would be impossible to impute a value and the total missing values only account for less than 0.01% of the total data.

In [31]:
# Checking for missing values in ords_prods_merge
ords_prods_merge.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
product_id                0
add_to_cart_order         0
reordered                 0
product_name              0
aisle_id                  0
department_id             0
prices                    0
_merge                    0
dtype: int64

In [32]:
# Checking for full duplicates in a dataframe
df_dups = ords_prods_merge[ords_prods_merge.duplicated()]

In [34]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, product_id, add_to_cart_order, reordered, product_name, aisle_id, department_id, prices, _merge]
Index: []

In [48]:
print(ords_prods_merge[ords_prods_merge['prices'] == 99999])

          order_id  user_id  order_number  orders_day_of_week  \
128911      183964      873             3                   0   
128924     1851256      873             4                   6   
284218     1915696     1893             1                   5   
284229     2763293     1893             2                   4   
284252     2564805     1893             4                   1   
...            ...      ...           ...                 ...   
32079919   2249946   204099            29                   0   
32079936   2363282   204099            31                   0   
32125745   3181945   204395            13                   3   
32255078   2486215   205227             7                   3   
32255132   2232716   205227            10                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
128911                   10                     7.0       33664   
128924                   12                    13.0       33664   
284218            

## 05. Exporting Data

In [35]:
# Exporting newly merged dataframe as a .pkl file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))

##### Decided to export as a .pkl file because of data set size.